File: librarymanager/librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/SectionEntity.java
package ru.elenacherev.librarymanager.domain.entity;

import lombok.*;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.util.HashSet;
import java.util.Set;

import static javax.persistence.GenerationType.IDENTITY;

@Entity
@Table(name = "SECTIONS")
@Getter
@NoArgsConstructor
//@RequiredArgsConstructor
@Setter
public class SectionEntity {
    @Id
    @GeneratedValue(strategy = IDENTITY) //уникальный ID, генерируется сам в БД
    @Column(name = "SECTION_ID")
    @NotNull
    @Setter(AccessLevel.NONE)
    private Long sectionId;

    @Version
    @Column(name = "VERSION")
    private int version; //версия, инкрементится при редактировании записи

    @NotEmpty(message = "{validation.title.notempty.message}")
    @Size(min = 3, max = 60, message = "{validation.title.size.message}")
    @Column(name = "TITLE")
    private String title; //

    @ManyToOne
    @JoinColumn(name = "PARENT_SECTION_ID")
    private SectionEntity parentSection;

    @OneToMany(mappedBy = "parentSection", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter(AccessLevel.NONE)
    private Set<SectionEntity> sections = new HashSet<SectionEntity>();
    
    @OneToMany(mappedBy = "section", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter(AccessLevel.NONE)
    private Set<EditionEntity> editions = new HashSet<EditionEntity>();
}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/SectionEntity.kt
package ru.elenacherev.librarymanager.domain.entity

import org.hibernate.annotations.ColumnDefault
import java.util.*
import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table
import javax.persistence.Version
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

@Entity
@Table(name = "SECTIONS")
data class SectionEntity(

    @Column(name = "SECTION_ID")
    @GeneratedValue
    @Id
    @get:NotNull
    var sectionId: UUID? = null,

    //версия, инкрементится при редактировании записи
    @Version
    @Column(name = "VERSION")
    var version: Int = 0,

    @Column(name = "TITLE")
    @NotEmpty
    @Size(min = 3, max = 60)
    var title: String,

    @ManyToOne
    @JoinColumn(name = "PARENT_SECTION_ID")
    var parentSection: SectionEntity? = null,

    @OneToMany(mappedBy = "parentSection", cascade = [CascadeType.ALL], orphanRemoval = true)
    val sections: MutableSet<SectionEntity> = mutableSetOf(),

    @OneToMany(mappedBy = "section", cascade = [CascadeType.ALL], orphanRemoval = true)
    val editions: MutableSet<EditionEntity> = mutableSetOf()
)